### PR TITLE
[monorepo] allow encoding array types

### PR DESCRIPTION
**Pull request by IIIIllllIIIIllllIIIIllllIIIIllllIIIIll**

### CODE REVIEW

File: packages/types/src/simple-types.ts [hunk 1]
```diff
@@ -10,22 +10,24 @@ export type Bytes32 = string;
 // around in protocol messages and include this in transaction data in disputes,
 // we impose some restrictions on the type; they must be serializable both as
 // JSON and as solidity structs.
-export type SolidityABIEncoderV2Struct = {
+export type SolidityABIEncoderV2Type = SolidityABIEncoderV2Struct | SolidityABIEncoderV2SArray;
```
Comment (cf19drofxots) [inline packages/types/src/simple-types.ts]: Is the union necessary here? Does the union with the last type for `SolidityABIEncoderV2Struct` not suffice?

Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [reply packages/types/src/simple-types.ts]: `SolidityABIEncoderV2Struct` disallows top-level arrays (e.g. the type `[1]` is not of type `SolidityABIEncoderV2Struct`) and unfortunately `tuple([1])` seems to be deserialized as `[1]`

Comment (snario) [reply packages/types/src/simple-types.ts]: I believe it is the difference between:

```solidity
struct State { bytes[] state; }
```

and

```solidity
bytes[] State;
```

Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [reply packages/types/src/simple-types.ts]: no, I mean that `struct State { bytes[] state; }` gets treated as `tuple(bytes[])` which gets treated as `bytes[]`

Comment (snario) [reply packages/types/src/simple-types.ts]: Ah.

Comment (IIIIllllIIIIllllIIIIllllIIIIllllIIIIll) [reply packages/types/src/simple-types.ts]: actually, now that I think about it, this sounds like a bug since we pass the `tuple` wrapper into ethers...

**Pull request merged**